File: fitnation/lib/models/MealPlan.dart
import 'package:uuid/uuid.dart';

class MealPlan {
  final String id;
  final String name;
  final String? description;
  final List<Meal> meals;
  final String? userGoals; // To connect with user's goals
  final String? linkedWorkoutPlanId; // To connect with workout plan

  MealPlan({
    required this.id,
    required this.name,
    this.description,
    required this.meals,
    this.userGoals,
    this.linkedWorkoutPlanId,
  });

  factory MealPlan.fromJson(Map<String, dynamic> json) {
    var mealsList = json['meals'] as List;
    List<Meal> meals = mealsList.map((i) => Meal.fromJson(i)).toList();

    return MealPlan(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      meals: meals,
      userGoals: json['user_goals'],
      linkedWorkoutPlanId: json['linked_workout_plan_id'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'meals': meals.map((meal) => meal.toJson()).toList(),
      'user_goals': userGoals,
      'linked_workout_plan_id': linkedWorkoutPlanId,
    };
  }
}

class Meal {
  final String name;
  final List<String> ingredients;
  final String instructions;
  final int? calories;
  final Map<String, dynamic>? macronutrients; // e.g., {'protein': 30, 'carbs': 40, 'fat': 15}

  Meal({
    required this.name,
    required this.ingredients,
    required this.instructions,
    this.calories,
    this.macronutrients,
  });

  factory Meal.fromJson(Map<String, dynamic> json) {
    return Meal(
      name: json['name'],
      ingredients: List<String>.from(json['ingredients']),
      instructions: json['instructions'],
      calories: json['calories'],
      macronutrients: json['macronutrients'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'ingredients': ingredients,
      'instructions': instructions,
      'calories': calories,
      'macronutrients': macronutrients,
    };
  }
}

File: backend/app/schemas/meal_plan.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime
import uuid

# Meal Schemas
class MealBase(BaseModel):
    name: str
    ingredients: List[str]
    instructions: str
    calories: Optional[int] = None
    macronutrients: Optional[Dict[str, Any]] = None # e.g., {'protein': 30, 'carbs': 40, 'fat': 15}

class MealCreate(MealBase):
    pass

class MealPublic(MealBase):
    # No ID for individual meals, as they are part of the meal plan's JSONB
    pass

# MealPlan Schemas
class MealPlanBase(BaseModel):
    name: str
    description: Optional[str] = None
    user_goals: Optional[str] = None
    linked_workout_plan_id: Optional[uuid.UUID] = None
    meals: List[MealCreate] # Use MealCreate for nested meals

class MealPlanCreate(MealPlanBase):
    pass

class MealPlanInDBBase(MealPlanBase):
    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

class MealPlanPublic(MealPlanInDBBase):
    # This will be the response model for GET requests
    # The 'meals' field is already included from MealPlanBase
    pass

class MealPlanFetch(MealPlanInDBBase):
    # This might be used for internal fetching if different from public
    pass
